File: frontend/src/api/customStrategy.ts
/**
 * 自定义策略API
 */

import apiClient from './client'
import type { AxiosResponse } from 'axios'

// 自定义策略信息
export interface CustomStrategyInfo {
  id: number
  user_id: number
  name: string
  description: string
  detailed_description: string
  parameter_descriptions: Record<string, string>
  is_public: boolean
  is_system: boolean
  created_at: string
  updated_at: string | null
}

// 自定义策略详情（包含代码）
export interface CustomStrategyDetail extends CustomStrategyInfo {
  code: string
}

// 创建策略请求
export interface CustomStrategyCreate {
  name: string
  description: string
  detailed_description: string
  code: string
  parameter_descriptions: Record<string, string>
}

// 更新策略请求
export interface CustomStrategyUpdate {
  name?: string
  description?: string
  detailed_description?: string
  code?: string
  parameter_descriptions?: Record<string, string>
}

// 策略列表响应
export interface CustomStrategyListResponse {
  strategies: CustomStrategyInfo[]
  total: number
}

// 策略验证请求
export interface CustomStrategyValidateRequest {
  code: string
  test_data?: boolean
}

// 策略验证响应
export interface CustomStrategyValidateResponse {
  valid: boolean
  errors: string[]
  warnings: string[]
  strategy_name?: string
  strategy_description?: string
}

export const customStrategyAPI = {
  // 获取策略列表
  getList(): Promise<AxiosResponse<CustomStrategyListResponse>> {
    return apiClient.get('/api/custom-strategy')
  },

  // 获取策略详情
  getDetail(strategyId: number): Promise<AxiosResponse<CustomStrategyDetail>> {
    return apiClient.get(`/api/custom-strategy/${strategyId}`)
  },

  // 创建策略
  create(data: CustomStrategyCreate): Promise<AxiosResponse<CustomStrategyInfo>> {
    return apiClient.post('/api/custom-strategy', data)
  },

  // 更新策略
  update(strategyId: number, data: CustomStrategyUpdate): Promise<AxiosResponse<CustomStrategyInfo>> {
    return apiClient.put(`/api/custom-strategy/${strategyId}`, data)
  },

  // 删除策略
  delete(strategyId: number): Promise<AxiosResponse<void>> {
    return apiClient.delete(`/api/custom-strategy/${strategyId}`)
  },

  // 验证策略代码
  validate(data: CustomStrategyValidateRequest): Promise<AxiosResponse<CustomStrategyValidateResponse>> {
    return apiClient.post('/api/custom-strategy/validate', data)
  },
}

File: backend/app/models/custom_strategy.py
"""自定义策略相关模型"""

from pydantic import BaseModel, Field
from typing import Optional, Dict, Any
from datetime import datetime


class CustomStrategyCreate(BaseModel):
    """创建自定义策略请求模型"""
    name: str = Field(..., description="策略名称", min_length=1, max_length=100)
    description: str = Field(default="", description="策略描述", max_length=500)
    detailed_description: str = Field(default="", description="详细说明")
    code: str = Field(..., description="策略代码", min_length=1)
    parameter_descriptions: Dict[str, str] = Field(default_factory=dict, description="参数说明")


class CustomStrategyUpdate(BaseModel):
    """更新自定义策略请求模型"""
    name: Optional[str] = Field(None, description="策略名称", min_length=1, max_length=100)
    description: Optional[str] = Field(None, description="策略描述", max_length=500)
    detailed_description: Optional[str] = Field(None, description="详细说明")
    code: Optional[str] = Field(None, description="策略代码", min_length=1)
    parameter_descriptions: Optional[Dict[str, str]] = Field(None, description="参数说明")


class CustomStrategyInfo(BaseModel):
    """自定义策略信息模型"""
    id: int
    user_id: int
    username: Optional[str] = None  # 创建人用户名
    name: str
    description: str
    detailed_description: str
    parameter_descriptions: Dict[str, str]
    is_public: bool
    is_system: bool
    created_at: str
    updated_at: Optional[str]
    
    class Config:
        from_attributes = True


class CustomStrategyDetail(CustomStrategyInfo):
    """自定义策略详情模型（包含代码）"""
    code: str


class CustomStrategyListResponse(BaseModel):
    """自定义策略列表响应模型"""
    strategies: list[CustomStrategyInfo]
    total: int


class CustomStrategyValidateRequest(BaseModel):
    """策略代码验证请求模型"""
    code: str = Field(..., description="策略代码", min_length=1)
    test_data: Optional[bool] = Field(default=False, description="是否使用测试数据验证")


class CustomStrategyValidateResponse(BaseModel):
    """策略代码验证响应模型"""
    valid: bool
    errors: list[str] = Field(default_factory=list, description="错误信息列表")
    warnings: list[str] = Field(default_factory=list, description="警告信息列表")
    strategy_name: Optional[str] = None
    strategy_description: Optional[str] = None
